### PR TITLE
[book-shop] added user entity and registration process

### PR DESCRIPTION
**Pull request by BolsunovaNataliia**

### CODE REVIEW

File: src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+@PasswordFieldMatch(message = "{register.repeatPassword.mismatch}")
+@Data
+public class UserRegistrationRequestDto {
+    private Long id;
+    @Email(message = "{register.email.invalid}")
+    @NotBlank(message = "{register.email.invalid}")
+    @NotNull
+    private String email;
+    @Length(min = 8, max = 35, message = "{register.password.size}")
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java]: Better don't mix jakarta bean validation annotations with hibernate specific annotations. Use Size where you need to validate text length.

File: src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import mate.academy.bookshop.validation.PasswordFieldMatch;
+import org.hibernate.validator.constraints.Length;
+
+@PasswordFieldMatch(message = "{register.repeatPassword.mismatch}")
+@Data
+public class UserRegistrationRequestDto {
+    private Long id;
+    @Email(message = "{register.email.invalid}")
+    @NotBlank(message = "{register.email.invalid}")
+    @NotNull
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java]: Do you need this annotation? Read what NotBlank annotation validates.

File: src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import jakarta.validation.constraints.Size;
+import lombok.Data;
+import mate.academy.bookshop.validation.PasswordFieldMatch;
+import org.hibernate.validator.constraints.Length;
+
+@PasswordFieldMatch(message = "{register.repeatPassword.mismatch}")
+@Data
+public class UserRegistrationRequestDto {
+    private Long id;
+    @Email(message = "{register.email.invalid}")
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java]: This annotation doesn't validate the length of input, so additionally to this annotation validate the input size  too.

File: src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    private String email;
+    @Length(min = 8, max = 35, message = "{register.password.size}")
+    @NotNull
+    private String password;
+    @Length(min = 8, max = 35)
+    @NotNull
+    private String repeatPassword;
+    @Size(max = 50)
+    @NotNull
+    private String firstName;
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java:28]: What is I pass an empty string here?

File: src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    @Size(max = 50)
+    @NotNull
+    private String lastName;
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java:31]: What is I pass an empty string here?

File: src/main/java/mate/academy/bookshop/exception/CustomGlobalExceptionHandler.java [hunk 6]
```diff
@@ -42,4 +43,11 @@ private String getErrorMessage(ObjectError e) {
         }
         return e.getDefaultMessage();
     }
+
+    @ExceptionHandler({RegistrationException.class})
+    public ResponseEntity<Object> handleRegistrationException(RegistrationException exception) {
+        return ResponseEntity
+                .status(HttpStatus.BAD_REQUEST)
+                .body(exception.getMessage());
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/exception/CustomGlobalExceptionHandler.java]: Pass a valid body (in JSON) in the same format your other handlers do.

File: src/main/java/mate/academy/bookshop/repository/user/UserRepository.java [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.util.Optional;
+import mate.academy.bookshop.model.User;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface UserRepository extends JpaRepository<User, Long>, JpaSpecificationExecutor<User> {
+    Optional<User> findByEmail(String email); // apache or util library?
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/repository/user/UserRepository.java]: What do you mean by this comment?

File: src/main/java/mate/academy/bookshop/service/UserServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto)
+            throws RegistrationException {
+        if (userRepository.findByEmail(requestDto.getEmail()).isPresent()) {
+            throw new RegistrationException("The user with this email is already registered "
+                    + requestDto.getEmail());
+        }
+        User user = new User();
+        user.setEmail(requestDto.getEmail());
```
Comment (vovasalyha) [inline src/main/java/mate/academy/bookshop/service/UserServiceImpl.java]: As I see you have a mapper for this.

**Review by vovasalyha — CHANGES_REQUESTED** [review]
Looks good, but see a couple of comments.

File: src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java (hunk 2 above)
Comment (bhdnchui) [reply src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java]: Let's return `@NotBlank` annotation here and remove `@NonNull`.
image

File: src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java (hunk 4 above)
Comment (bhdnchui) [reply src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java:28]: `@NotBlank` is better to use, when you process strings.
Also, pay attention that `@NotNull` is redundant if there is `@NotBlank` or `@NotEmpty`.

File: src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java (hunk 5 above)
Comment (bhdnchui) [reply src/main/java/mate/academy/bookshop/dto/UserRegistrationRequestDto.java:31]: Same

File: src/main/java/mate/academy/bookshop/validation/PasswordMatchValidator.java [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+        ConstraintValidator<PasswordFieldMatch, UserRegistrationRequestDto> {
+    @Override
+    public void initialize(PasswordFieldMatch p) {
+    }
+
+    public boolean isValid(UserRegistrationRequestDto requestDto, ConstraintValidatorContext c) {
+        String password = requestDto.getPassword();
+        String repeatPassword = requestDto.getRepeatPassword();
+
+        if (password == null || !password.equals(repeatPassword)) {
```
Comment (bhdnchui) [inline src/main/java/mate/academy/bookshop/validation/PasswordMatchValidator.java]: Let's think about using Objects#equals here.

**Review by bhdnchui — CHANGES_REQUESTED** [review]
Good job!
Just a few comments to process.

**Review by bhdnchui — APPROVED** [review]
LGTM

**Pull request merged**